### DISCUSSION TITLE
The Right way to use Styles

### DISCUSSION THREAD

**Discussion by LSXAxeller** [AvaloniaUI/Avalonia#15864] [Q&A] [answered]
I want to make some buttons totally transparent (except for the icon) even on hover and press. So, I tried using something like this
```
<Button Name="PART_TogglePaneBtn" Grid.Column="0" VerticalAlignment="Top"
                                        Height="50"
                                        Width="50" 
                                        Styles="{StaticResource ButtonTransparent}"
                                        Margin="10,0,0,0">
                                    <icons:SymbolIcon Symbol="Options" FontSize="26" />
                                </Button>
```

However, I am encountering a build error: `  SessionsListControl.axaml(40, 41): [AVLN3000] Unable to find suitable setter or adder for property Styles of type Avalonia.Base:Avalonia.StyledElement for argument System.Runtime:System.Object, available setter parameter lists are:
Avalonia.Styling.IStyle
Avalonia.Styling.IStyle Line 40, position 41.` on line `Styles="{StaticResource ButtonTransparent}"`

this is what I did

```
<Application xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:icons="clr-namespace:Material.Icons.Avalonia;assembly=Material.Icons.Avalonia"
             xmlns:fluent="using:FluentAvalonia.Styling"
             RequestedThemeVariant="Dark">

    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <ResourceInclude Source="Styles/ButtonTransparent.axaml"/>
            </ResourceDictionary.MergedDictionaries>
        </ResourceDictionary>
    </Application.Resources>

    <Application.Styles>
        <icons:MaterialIconStyles />
        <fluent:FluentAvaloniaTheme />
    </Application.Styles>
</Application>
```

```
<ResourceDictionary xmlns="https://github.com/avaloniaui"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
    <Styles x:Key="ButtonTransparent" >
        <Style Selector="Button">
            <Setter Property="Background" Value="Transparent" />
            <Setter Property="BorderThickness" Value="0" />
            <Setter Property="Padding" Value="0" />
            <Setter Property="Template">
                <ControlTemplate>
                    <Grid>
                        <ContentPresenter Content="{TemplateBinding Content}" />
                    </Grid>
                </ControlTemplate>
            </Setter>
        </Style>
        <Style Selector="Button:pointerover /template/ContentPresenter">
            <Setter Property="Background" Value="Transparent" />
            <Setter Property="BorderThickness" Value="0" />
            <Setter Property="Padding" Value="0" />
        </Style>
        <Style Selector="Button:pressed /template/ContentPresenter">
            <Setter Property="Background" Value="Transparent" />
            <Setter Property="BorderThickness" Value="0" />
            <Setter Property="Padding" Value="0" />
        </Style>
    </Styles>
</ResourceDictionary>
```

**Comment by stevemonaco** [comment, marked as answer]
The idiomatic ways are to:
1. Use class selectors. ie. `<Style Selector="Button.transparent">` added / `<StyleInclude />` into a `Styles` collection. Then `<Button Classes="transparent" />`
2. Create a new `ControlTheme`, merge it, and set `<Button Theme="{StaticResource YourControlTheme}" />`

I'd probably use class selectors.

**Reply by LSXAxeller** [reply]
thanks, using class selectors worked perfectly.

btw, is there any good sources where I can learn more about styling classes? The main documentation isn't very helpful in that aspect and youtube and udemy lacks content about avalonia in general

**Reply by stevemonaco** [reply]
Style classes to do what exactly? Customize existing controls? Style custom controls? The docs are comprehensive regarding what selectors are available https://docs.avaloniaui.net/docs/reference/styles/style-selector-syntax

Beyond that, it's not really worth publishing specific content on: there's simply too many things that can be customized. The proper mindset is to know a general approach. Namely, you need to be reading the original `ControlTheme` for the specific theme you're using for anything you customize. Otherwise, it's like using a new API without learning anything about it and not having IntelliSense either. `ControlTheme`s define the entire look-and-feel of the control and very few pieces are required. Additional pieces are at the discretion of the theme implementer.
